File: api/src/models/activity.py
import uuid
from datetime import date, datetime, time
from decimal import Decimal
from typing import TYPE_CHECKING
from uuid import UUID as _UUID

from sqlalchemy import Boolean, Date, DateTime, ForeignKey, Numeric, String, Time
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import Mapped, mapped_column, relationship
from sqlalchemy.sql import func

from src.config.database import Base

if TYPE_CHECKING:
    from src.models.trip import Trip


class Activity(Base):
    __tablename__ = "activities"

    id: Mapped[_UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    trip_id: Mapped[_UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("trips.id"), nullable=False, index=True
    )
    title: Mapped[str] = mapped_column(String, nullable=False)
    description: Mapped[str | None] = mapped_column(String, nullable=True)
    # Nullable since SMP-324: AI-generated MEAL / TRANSPORT recommendations
    # are not pinned to a calendar slot — they live in dedicated sections
    # of the review screen and the trip detail. Dated itinerary entries
    # (CULTURE, NATURE, ...) still use this column as before.
    date: Mapped[date | None] = mapped_column(Date, nullable=True)
    start_time: Mapped[time | None] = mapped_column(Time, nullable=True)
    end_time: Mapped[time | None] = mapped_column(Time, nullable=True)
    location: Mapped[str | None] = mapped_column(String, nullable=True)
    category: Mapped[str] = mapped_column(String, nullable=False, default="OTHER")
    estimated_cost: Mapped[Decimal | None] = mapped_column(Numeric(12, 2), nullable=True)
    is_booked: Mapped[bool] = mapped_column(Boolean, nullable=False, default=False)
    is_done: Mapped[bool] = mapped_column(
        Boolean, nullable=False, default=False, server_default="false"
    )
    validation_status: Mapped[str] = mapped_column(
        String, nullable=False, default="MANUAL", server_default="MANUAL"
    )
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now())
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), onupdate=func.now()
    )

    trip: Mapped["Trip"] = relationship("Trip", back_populates="activities")

File: bagtrip/lib/models/activity.dart
import 'package:freezed_annotation/freezed_annotation.dart';

part 'activity.freezed.dart';
part 'activity.g.dart';

@JsonEnum(alwaysCreate: true)
enum ActivityCategory {
  @JsonValue('VISIT')
  visit,
  @JsonValue('RESTAURANT')
  restaurant,
  @JsonValue('TRANSPORT')
  transport,
  @JsonValue('LEISURE')
  leisure,
  @JsonValue('CULTURE')
  culture,
  @JsonValue('NATURE')
  nature,
  @JsonValue('OTHER')
  other,
}

@JsonEnum(alwaysCreate: true)
enum ValidationStatus {
  @JsonValue('SUGGESTED')
  suggested,
  @JsonValue('VALIDATED')
  validated,
  @JsonValue('MANUAL')
  manual,
}

@freezed
abstract class Activity with _$Activity {
  const factory Activity({
    required String id,
    @JsonKey(name: 'trip_id') required String tripId,
    required String title,
    String? description,
    required DateTime date,
    @JsonKey(name: 'start_time') String? startTime,
    @JsonKey(name: 'end_time') String? endTime,
    String? location,
    @JsonKey(unknownEnumValue: ActivityCategory.other)
    @Default(ActivityCategory.other)
    ActivityCategory category,
    @JsonKey(name: 'estimated_cost') double? estimatedCost,
    @JsonKey(name: 'is_booked') @Default(false) bool isBooked,
    @JsonKey(name: 'validation_status')
    @Default(ValidationStatus.manual)
    ValidationStatus validationStatus,
    int? suggestedDay,
    @JsonKey(name: 'created_at') DateTime? createdAt,
    @JsonKey(name: 'updated_at') DateTime? updatedAt,
  }) = _Activity;

  factory Activity.fromJson(Map<String, dynamic> json) =>
      _$ActivityFromJson(json);
}
